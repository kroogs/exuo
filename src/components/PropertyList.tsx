/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 * Copyright Â© 2020 Justin Krueger */

import React from 'react'

import { IAnyStateTreeNode, getMembers } from 'mobx-state-tree'
import { useObserver } from 'mobx-react-lite'

import { createStyles, makeStyles, Theme } from '@material-ui/core/styles'
import List from '@material-ui/core/List'
import ListItem from '@material-ui/core/ListItem'
import ListItemText from '@material-ui/core/ListItemText'

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      width: '100%',
      backgroundColor: theme.palette.background.default,
    },
  }),
)

interface PropertyListProps {
  model: IAnyStateTreeNode | null
  options?: Partial<{
    groupBy: null
  }>
}

export const PropertyList: React.FunctionComponent<PropertyListProps> = props => {
  const classes = useStyles()

  return useObserver(() => {
    if (!props.model) {
      throw Error('Invalid model')
    }

    const members = getMembers(props.model)
    const properties = Object.keys(members.properties)

    return (
      <>
        <span>Property List</span>
        <List className={classes.root}>
          {properties.map(value => (
            <ListItem key={value} role={undefined} dense button>
              <ListItemText id={`list-label-${value}`} primary={value} />
            </ListItem>
          ))}
        </List>
      </>
    )
  })
}

export default PropertyList
