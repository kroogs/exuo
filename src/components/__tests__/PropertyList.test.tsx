/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 * Copyright Â© 2020 Justin Krueger */

import React from 'react'
import { shallow } from 'enzyme'
import { types as t } from 'mobx-state-tree'

import PropertyList from '../PropertyList'

test('Renders a list of properties', () => {
  const Model = t.model({
    one: t.number,
    two: t.string,
    three: t.boolean,
  })

  const instance = Model.create({
    one: 100,
    two: 'graphs',
    three: true,
  })

  expect(shallow(<PropertyList model={instance} />)).toMatchSnapshot()
})
