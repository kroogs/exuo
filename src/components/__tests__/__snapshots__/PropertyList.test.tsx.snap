// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders a list of properties 1`] = `
<WithStyles(ForwardRef(List))
  className="makeStyles-root-1"
>
  <WithStyles(ForwardRef(ListItem))
    button={true}
    dense={true}
    key="one"
  >
    <WithStyles(ForwardRef(ListItemText))
      id="list-label-one"
      primary="one"
    />
  </WithStyles(ForwardRef(ListItem))>
  <WithStyles(ForwardRef(ListItem))
    button={true}
    dense={true}
    key="two"
  >
    <WithStyles(ForwardRef(ListItemText))
      id="list-label-two"
      primary="two"
    />
  </WithStyles(ForwardRef(ListItem))>
  <WithStyles(ForwardRef(ListItem))
    button={true}
    dense={true}
    key="three"
  >
    <WithStyles(ForwardRef(ListItemText))
      id="list-label-three"
      primary="three"
    />
  </WithStyles(ForwardRef(ListItem))>
</WithStyles(ForwardRef(List))>
`;
